userspace/libsinsp/container_engine/bpm.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright (C) 2013-2019 Draios Inc dba Sysdig.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+#include "container_engine/bpm.h"
+#include "sinsp.h"
+
+using namespace libsinsp::container_engine;
+
+bool bpm::resolve(sinsp_container_manager* manager, sinsp_threadinfo* tinfo, bool query_os_for_missing_info)
+{
+	sinsp_container_info container_info;
+	bool matches = false;
+
+	for(auto it = tinfo->m_cgroups.begin(); it != tinfo->m_cgroups.end(); ++it)
+	{
+		string cgroup = it->second;
+		size_t pos;
+
+		//
+		// Non-systemd and systemd BPM
+		//
+		pos = cgroup.find("bpm-");
+		if(pos != string::npos)
+		{
+			auto id_start = pos + sizeof("bpm-") - 1;
+			auto id_end = cgroup.find(".scope", id_start);
```

Comment:
The BPM container ID format is defined [here](https://github.com/cloudfoundry-incubator/bpm-release/blob/master/src/bpm/jobid/encoding.go).
The format of BPM v1.0.3 is a combination of `abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789` and `_-`. All other characters get escaped by a `.` followed by 2 hex digits.

Example IDs can be listed by `/var/vcap/packages/bpm/bin/runc --root /var/vcap/data/bpm/runc/ list --quiet` on a BOSH VM:
```
bpm-blobstore
bpm-director
bpm-director.2enginx
bpm-director.2escheduler
bpm-director.2esync_dns
bpm-nats
bpm-postgres-10
```
Note how `.` got encoded into `.2e`.

Therefore, I guess there is no need to truncate them to 12 hex digits?
Also, there is no need to ensure that the container ID between `bpm-` and `.scope` contains slashes because they are hex encoded.

Reply:
Not truncating is perfectly fine. You probably want to ensure there are no characters outside `[a-zA-Z0-9._-]` before (or while) you decode the string though.

Reply:
done